File: 977. Squares of a Sorted Array/SortedSquares.py
class Solution:
    def sortedSquares(self, nums: list[int]):
        # Trivial solution/ O(nlogn)
        ans = [(num**2) for num in nums]
        ans.sort()
        
        return ans

    def sortedSquares(self, nums: list[int]):
        # O(n) solution
        n = len(nums)
        l, r = None, None
        for index, num in enumerate(nums):
            if num >= 0:
                r = index
                l = index - 1
                break

        if l and l < 0:   return [(num**2) for num in nums]
        if r == None:   return [(nums[i] **2) for i in range(n-1, -1, -1)]
        
        ans = []
        while l >= 0 and r < n:
            if abs(nums[l]) < nums[r]:
                ans.append(nums[l]** 2)
                l -=1
            else:
                ans.append(nums[r]**2)
                r +=1

        while l >= 0:
            ans.append(nums[l] **2)
            l -=1
        while r < n:
            ans.append(nums[r] **2)
            r +=1

        return ans

File: 977. Squares of a Sorted Array/sorted_squares.rs
impl Solution {
    
    // O(n) time,
    // O(n) space,
    // Approach: two pointers, 
    pub fn sorted_squares(nums: Vec<i32>) -> Vec<i32> {        
        let (mut left, mut right) = (0, 0);
        for (i, num) in nums.iter().enumerate() {
            if *num >= 0 {
                break;
            }
            right = i+1;
        }
        left = right-1;
        
        let mut answer = vec![];
        while (left >= 0 && left < nums.len()) && right < nums.len() {
            let left_sqr = nums[left]*nums[left];
            let right_sqr = nums[right]*nums[right];
            if left_sqr < right_sqr {
                answer.push(left_sqr);
                left -= 1;
            } else {
                answer.push(right_sqr);
                right += 1;
            }
        }
        
        while left >= 0 && left < nums.len() {
            answer.push(nums[left]*nums[left]);
            left -= 1;
        }
        
        while right < nums.len() {
            answer.push(nums[right]*nums[right]);
            right += 1;
        }
        
        answer
    }
}